### PR TITLE
USB replication: Add grub entries, improve set image handling

### PR DESCRIPTION
**Pull request by chaitu236**
This patch set has rudimentary implementation for Get/Set Image Grub menu options in provisioning media.

New grub menu entries:

image

`Get Image` entry output:

image

`Set Image` entry output:

image

Patch set has following changes

- Add grub menu entries for Get/Set Image
- Use `bootdelay` and `consoleoutputenable` from image but initialize other grubenv entries to defaults.
- Add missing dependencies of ni-netcfgutil.
- Handle netcfg operations in setall.
- Fix populateconfig so it works in restoremode.

WI: [AB#2372259](https://dev.azure.com/ni/DevCentral/_workitems/edit/2372259)

### Testing
- [x] Tested "Get Image" grub entry on cRIO-9030 and a VM
- [x] Tested "Set Image" grub entry on cRIO-9030 with image from the same target and image from a VM
- [x] hostname, ssh keys are handled correctly i.e., they're not replicated

### Known issues
- Network configuration like static ip address setting are also being replicated which is probably not what we want. Will address this in another PR.
- Error handling (e.g., when media is read-only, etc), ability to provide image name, ability to pick a certain image will be addressed in another PR.

### CODE REVIEW

File: recipes-ni/ni-systemreplication/files/nisystemreplication [hunk 1]
```diff
@@ -12,12 +12,34 @@ image_get () {
...
+	echo $LOG_LEVEL > /proc/sys/kernel/printk
+	sanity_check
+
+	print_info "Re-enabling automount..."
+	enable_automount
+	print_done
+
+	trap - ERR
+	exec 1>&3
+	exec 2>&4
```
Comment (amstewart) [inline recipes-ni/ni-systemreplication/files/nisystemreplication:52]: What are these for?

File: recipes-ni/ni-systemreplication/files/nisystemreplication [hunk 2]
```diff
@@ -12,12 +12,34 @@ image_get () {
...
+
 	nisystemimage setall -x tgz -f getall.tgz -p reset -s reset
 
-	# save rootfs UUID for grub to reference
-	ROOTUUID=`lsblk ${TARGET_DISK}${PART_SEPARATOR}4 -n -o PARTUUID`
-	echo set rootuuid=$ROOTUUID >> $GRUB_MOUNTPOINT/grubvar_readonly
+	# Retain some grubenv settings
+	if grep --quiet "^consoleoutenable=" $BOOTFS_MOUNTPOINT/grub/grubenv; then
+		grubenv_consoleoutenable=`grub-editenv $BOOTFS_MOUNTPOINT/grub/grubenv list |grep "^consoleoutenable=" | cut -f 2 -d '='`
+	fi
```
Comment (amstewart) [inline recipes-ni/ni-systemreplication/files/nisystemreplication]: It isn't clear to me were these variables are getting used. If they're being used by the `install_grubenv` func below, can they just be calculated there?

File: recipes-ni/ni-systemreplication/files/nisystemreplication [hunk 3]
```diff
@@ -12,12 +12,34 @@ image_get () {
...
+	if grep --quiet "^bootdelay=" $BOOTFS_MOUNTPOINT/grub/grubenv; then
+		grubenv_bootdelay=`grub-editenv $BOOTFS_MOUNTPOINT/grub/grubenv list |grep "^bootdelay=" | cut -f 2 -d '='`
```
Comment (amstewart) [inline recipes-ni/ni-systemreplication/files/nisystemreplication]: NIT: Prefer `$(...)` to backtick styling.

File: recipes-ni/ni-systemreplication/files/nisystemreplication [hunk 4]
```diff
@@ -5,18 +5,28 @@ set -e
 # get or set
 declare -r MODE="${1:-}"
 
+NIRECOVERY_MOUNTPOINT=/mnt/NIRECOVERY
+
 image_get () {
-	nisystemimage getall -x tgz -f getall.tgz
+	mount -o remount,rw $NIRECOVERY_MOUNTPOINT
+	mkdir -p $NIRECOVERY_MOUNTPOINT/replication_images
+	nisystemimage getall -x tgz -f $NIRECOVERY_MOUNTPOINT/replication_images/getall.tgz
```
Comment (amstewart) [inline recipes-ni/ni-systemreplication/files/nisystemreplication]: MINOR: Can we figure out a more meaningful name for the archive than `getall.tgz`? Maybe: `latest.tgz`?

File: scripts/lib/wic/canned-wks/grub_old.cfg [hunk 5]
```diff
@@ -1,6 +1,6 @@
 serial --unit=0 --speed=115200 --word=8 --parity=no --stop=1
 default='Default Provisioning'
-timeout=1
+timeout=3
```
Comment (amstewart) [inline scripts/lib/wic/canned-wks/grub_old.cfg]: What's the reason for extending the timeout? Just to make it easier to select the new options?

File: recipes-ni/ni-systemimage/files/nisystemimage [hunk 6]
```diff
@@ -634,7 +641,32 @@ init_fd () {
 	init_fd_$MODE
 }
 
+init_fs_restoremode () {
+	# If the util did not previously exit cleanly, it's possible that the
+	# working mount points didn't get unmount them. To avoid double-mounts,
```
Comment (amstewart) [inline recipes-ni/ni-systemimage/files/nisystemimage]: Could you fixup the grammar in this comment you copied.

File: recipes-ni/ni-systemimage/files/nisystemimage [hunk 7]
```diff
@@ -634,7 +641,32 @@ init_fd () {
...
+	run_cmd_always mount -L $BOOTFS_LABEL $ORIG_BOOTFS \
+		|| die FS_PREP_FAILED "Couldn't mount label $BOOTFS_LABEL to $ORIG_BOOTFS "
+	run_cmd_always mkdir -p $ORIG_CONFIGFS
+	run_cmd_always mount -L $CONFIGFS_LABEL $ORIG_CONFIGFS \
+		|| die FS_PREP_FAILED "Couldn't mount label $CONFIGFS_LABEL to $ORIG_CONFIGFS"
+}
+
 init_fs () {
+	if in_restoremode; then
+		# Tools such as ninetcfg and nirtcfg same mountpoints as safemode and runmode
```
Comment (amstewart) [inline recipes-ni/ni-systemimage/files/nisystemimage]: Grammar.

**Review by amstewart — CHANGES_REQUESTED** [review]
Since you have the context now, can we figure out a better name for the `grub_old.cfg` OE file?

File: recipes-ni/ni-systemreplication/files/nisystemreplication (hunk 4 above)
Comment (chaitu236) [reply recipes-ni/ni-systemreplication/files/nisystemreplication]: `latest.tgz` may also cause confusion about what kind of image that is - the one containing just runmode and config (i.e., get operation) or the one containing grub, safemode, runmode, config (i.e., getall).

How about `systemimage.tgz` or `latest_systemimage.tgz`?

File: scripts/lib/wic/canned-wks/grub_old.cfg (hunk 5 above)
Comment (chaitu236) [reply scripts/lib/wic/canned-wks/grub_old.cfg]: Yes, 1s was too short to be able to select new options.

File: recipes-ni/ni-systemreplication/files/nisystemreplication (hunk 2 above)
Comment (chaitu236) [reply recipes-ni/ni-systemreplication/files/nisystemreplication]: They're used in `install_grubenv` [here](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.safemode.common#L304) and [here](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.safemode.common#L322).

They were originally added to be set from [early_setup](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.common#L502), but we can't determine the values for these until the `getall.tgz` archive is unpacked.. so we have to do this after `nisystemimage setall`.

> can they just be calculated there

`install_grubenv` needs to be able to work as before for regular provisioning.. can't think of a good way to move it into that script.

File: recipes-ni/ni-systemreplication/files/nisystemreplication (hunk 4 above)
Comment (amstewart) [reply recipes-ni/ni-systemreplication/files/nisystemreplication]: I'd be fine with either. I suppose the former makes more sense if this workflow isn't supposed to support having multiple images on the tool at once.

File: scripts/lib/wic/canned-wks/grub_old.cfg (hunk 5 above)
Comment (amstewart) [reply scripts/lib/wic/canned-wks/grub_old.cfg]: OK. Could you just add that reasoning to the commit message. It's the kind of logic that someone is going to have to check in the git logs in the future.

File: recipes-ni/ni-systemreplication/files/nisystemreplication (hunk 1 above)
Comment (chaitu236) [reply recipes-ni/ni-systemreplication/files/nisystemreplication:52]: Copied these from the regular provisioning codeflow https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.safemode#L20.

I believe they're undoing the [custom error handler](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.common#L460) and IO redirection done earlier in [early_setup](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.common#L479).

**Comment by chaitu236** [comment]
> Since you have the context now, can we figure out a better name for the `grub_old.cfg` OE file?

Moved `grub_old.cfg` to `grub.cfg` as discussed.

File: recipes-ni/ni-systemreplication/files/nisystemreplication (hunk 4 above)
Comment (chaitu236) [reply recipes-ni/ni-systemreplication/files/nisystemreplication]: Done. It's just one image for now but we'll expand it to multiple.

File: scripts/lib/wic/canned-wks/grub_old.cfg (hunk 5 above)
Comment (chaitu236) [reply scripts/lib/wic/canned-wks/grub_old.cfg]: Done

File: recipes-ni/ni-systemimage/files/nisystemimage (hunk 6 above)
Comment (chaitu236) [reply recipes-ni/ni-systemimage/files/nisystemimage]: Done

File: recipes-ni/ni-systemimage/files/nisystemimage (hunk 7 above)
Comment (chaitu236) [reply recipes-ni/ni-systemimage/files/nisystemimage]: Done

File: recipes-ni/ni-systemreplication/files/nisystemreplication (hunk 3 above)
Comment (chaitu236) [reply recipes-ni/ni-systemreplication/files/nisystemreplication]: Done

**Review by amstewart — APPROVED** [review]

**Pull request merged**